File: project_euler/Problem004/Problem004.java
package project_euler;

public class Problem004 {
    public void solution() {
        int target = 100;
        int result = 0;
        for (int i = 999; i >= target; --i) {
            for (int j = i; j >= target; --j) {
                int value = i * j;
                if (isPalindrome(value) && result < value) {
                    result = value;
                    target = j;
                    break;
                }
            }
        }
        System.out.println(result);
    }

    private boolean isPalindrome(int value) {
        int divisor = 1;
        while (10 <= value / divisor) {
            divisor *= 10;
        }

        while (value > 1) {
            int head = value / divisor;
            int tail = value % 10;
            if (head != tail) {
                return false;
            }

            value = value % divisor / 10;
            divisor /= 100;
        }
        return true;
    }
}

File: project_euler/Problem004/Problem004.go
package project_euler

import "fmt"

func Problem004() {
    target := 100
    result := 0
    for i := 999; i >= target; i-- {
        for j := i; j >= target; j-- {
            value := i * j
            if isPalindrome(value) && result < value {
                result = value
                target = j
                break
            }
        }
    }
    fmt.Println(result)
}

func isPalindrome(value int) bool {
    divisor := 1
    for 10 <= value / divisor {
        divisor *= 10
    }

    for value > 1 {
        head := value / divisor
        tail := value % 10
        if head != tail {
            return false
        }

        value = value % divisor / 10
        divisor /= 100
    }
    return true
}
